File: server/datasets/python_source-code/Multiple-C2/organic_8.py
"""
0/1 Knapsack (DP) - Organic Submission #8
Functional style using a dataclass for items and full type hints.
"""

from dataclasses import dataclass
from typing import List


@dataclass(frozen=True)
class Item:
    weight: int
    value: int


def max_value(items: List[Item], capacity: int) -> int:
    dp: List[int] = [0] * (capacity + 1)

    for item in items:
        for cap in range(capacity, item.weight - 1, -1):
            candidate = dp[cap - item.weight] + item.value
            if candidate > dp[cap]:
                dp[cap] = candidate

    return dp[capacity]


def main() -> None:
    items = [Item(2, 3), Item(3, 4), Item(4, 5), Item(5, 6)]
    print(max_value(items, 5))


if __name__ == "__main__":
    main()

File: server/datasets/java_source-code/Multiple-C2/organic_8.java
/*
 * 0/1 Knapsack (DP) - Organic Submission #8
 * Functional style using Item object and 1D DP table.
 */

import java.util.ArrayList;
import java.util.List;

class Item {
    int weight;
    int value;

    public Item(int weight, int value) {
        this.weight = weight;
        this.value = value;
    }
}

class Solution {
    public static int maxValue(List<Item> items, int capacity) {
        int[] dp = new int[capacity + 1];

        for (Item item : items) {
            for (int cap = capacity; cap >= item.weight; cap--) {
                int candidate = dp[cap - item.weight] + item.value;
                if (candidate > dp[cap]) {
                    dp[cap] = candidate;
                }
            }
        }
        return dp[capacity];
    }

    public static void main(String[] args) {
        List<Item> items = new ArrayList<>();
        items.add(new Item(2, 3));
        items.add(new Item(3, 4));
        items.add(new Item(4, 5));
        items.add(new Item(5, 6));
        System.out.println(maxValue(items, 5));
    }
}
